mongoengine/fields.py

```diff
@@ -863,6 +864,14 @@ def validate(self, value):
         """Make sure that a list of valid fields is being used."""
         if not isinstance(value, (list, tuple, BaseQuerySet)):
             self.error('Only lists and tuples may be used in a list field')
+
+        # Validate that max_length is not exceeded.
+        # NOTE It's still possible to bypass this enforcement by using $push.
+        # However, if the document is reloaded after $push and then re-saved,
+        # the validation error will be raised.
+        if self.max_length is not None and len(value) > self.max_length:
```

Comment:
Shouldn't we also implement the check from our fork where `prepare_query_value` ensures you can't go over the limit using $set (and add a test for this scenario)? https://github.com/closeio/mongoengine/blob/master/mongoengine/fields.py#L602

Reply:
Good catch, thanks!

Added a test case (which our fork was missing) and the relevant `prepare_query_value` code in 87194856.